### PR TITLE
Add codgen tests

### PR DESCRIPTION
**Pull request by sugmanue**
## Motivation and Context

Adds codegen tests for RPCv2, for:

* Client
* Async client
* Marshalling transformers.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/rpcv2/customization.config [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (zoewangg) [inline codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/rpcv2/customization.config:1]: Nit: do we need customization.config and endpoint-rule-set files?

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java [hunk 2]
```diff
@@ -55,7 +55,7 @@ public abstract class BaseAwsJsonProtocolFactory {
     /**
      * Used by operations that do not serialize the input, e.g., when the input is not defined in the model. RPCv2 uses it.
      */
-    public static final OperationMetadataAttribute<Boolean> USE_NO_OP_GENERATOR = new OperationMetadataAttribute<>(Boolean.class);
+    public static final OperationMetadataAttribute<Boolean> GENERATES_BODY = new OperationMetadataAttribute<>(Boolean.class);
```
Comment (zoewangg) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java:58]: This is protected API, so we should probably honor backwards compatibility

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.protocols.json;
```
Comment (zoewangg) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java:16]: Internal APIs should be under internal subpackage

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.Map;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.protocols.core.OperationInfo;
+import software.amazon.awssdk.utils.MapUtils;
+
+/**
+ * Records facts about each protocol.
+ */
+@SdkInternalApi
+public enum ProtocolFact {
```
Comment (zoewangg) [inline core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java:28]: Should we use interface instead?

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java (hunk 2 above)
Comment (sugmanue) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/BaseAwsJsonProtocolFactory.java:58]: > This is protected API, so we should probably honor backwards compatibility

This attribute has not yet been published and was introduced in the past PR. Just changing the name to clarify its intent.

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java (hunk 4 above)
Comment (sugmanue) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java:28]: We could but I like the enum just fine. Let me know if you feel strong about it and I can change it.

File: codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/rpcv2/customization.config (hunk 1 above)
Comment (sugmanue) [reply codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/c2j/rpcv2/customization.config:1]: removed.

File: core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java (hunk 4 above)
Comment (zoewangg) [reply core/protocols/aws-json-protocol/src/main/java/software/amazon/awssdk/protocols/json/ProtocolFact.java]: Is there any benefit of using enum, it seems a bit unconventional to me to use enum for behavior

**Review by zoewangg — APPROVED** [review]

**Pull request merged**